### PR TITLE
🌱 maintainer annotations: search for config

### PR DESCRIPTION
**Pull request by raghavkaul**
#### What kind of change does this PR introduce?
Look for maintainer annotation config files in multiple locations: `scorecard.yml`, `.scorecard.yml`, `.github/scorecard.yml`. Scorecard uses the first one it finds regardless of whether it's valid.

#### What is the current behavior?
Only look for `scorecard.yml` and don't apply maintainer annotation if that file doesn't exist.

- [x] Tests for the changes have been added (for bug fixes/features)

#### Which issue(s) this PR fixes
Addresses #4048

### CODE REVIEW

File: pkg/scorecard.go [hunk 1]
```diff
@@ -168,21 +169,23 @@ func runScorecard(ctx context.Context,
...
 			}
-			c, err := config.Parse(rc, checks)
+			c, err := config.Parse(r, checks)
 			if err != nil {
-				logger := sclog.NewLogger(sclog.DefaultLevel)
 				logger.Error(err, "parsing configuration file")
 			}
 			ret.Config = c
+		} else {
+			logger.Info(fmt.Sprintf("no maintainer annotation: %v", err))
```
Comment (spencerschrock) [inline pkg/scorecard.go:188]: I lean towards the opposite. Logging if a maintainer annotation file is found and used, not if it's absent.

For example, golangci-lint spits this out, and we could do the same:
```
INFO [config_reader] Used config file .golangci.yml
```

Comment (spencerschrock) [reply pkg/scorecard.go:188]: Plus, I don't think `err` is set appropriately there and shouldn't be in the message.

File: pkg/scorecard_test.go [hunk 2]
```diff
@@ -341,3 +346,65 @@ func TestExperimentalRunProbes(t *testing.T) {
...
+		{
+			desc:      ".github/scorecard.yml exists",
+			locs:      []string{".github/scorecard.yml"},
+			wantFound: true,
+		},
+		{
+			desc:      "multiple configs exist",
+			locs:      []string{"scorecard.yml", ".github/scorecard.yml"},
+			wantFound: true,
+		},
```
Comment (spencerschrock) [inline pkg/scorecard_test.go:378]: do we want to test precedence at all? or nah?

File: pkg/scorecard_test.go [hunk 3]
```diff
@@ -341,3 +346,65 @@ func TestExperimentalRunProbes(t *testing.T) {
...
+			mockRepoClient.EXPECT().GetFileReader(gomock.Any()).AnyTimes().DoAndReturn(func(filename string) (io.ReadCloser, error) {
+				if !slices.Contains(tt.locs, filename) {
+					return nil, fmt.Errorf("os.Open: %s", filename)
+				}
+				fullPath := filepath.Join("./testdata", filename)
+				f, err := os.Open(fullPath)
+				if err != nil {
+					return nil, fmt.Errorf("os.Open: %w", err)
+				}
+				return f, nil
```
Comment (spencerschrock) [inline pkg/scorecard_test.go:401]: I don't think the slices.Contains block is needed. If the file doesn't exist, os.Open will fail anyway.

So something like  this should be sufficient (wrapping linter is disabled for test funcs)
```go
mockRepoClient.EXPECT().GetFileReader(gomock.Any()).AnyTimes().DoAndReturn(func(filename string) (io.ReadCloser, error) {
	fullPath := filepath.Join("./testdata", filename)
	return os.Open(fullPath)
})
```

**Review by spencerschrock — APPROVED** [review]
Seems reasonable. Let me know what you think about the logging flip. And please document the config precedence in `config/README.md`

File: pkg/scorecard_test.go (hunk 2 above)
Comment (raghavkaul) [reply pkg/scorecard_test.go]: Added a test for precedence

File: pkg/scorecard_test.go (hunk 3 above)
Comment (raghavkaul) [reply pkg/scorecard_test.go:406]: This is a bit confusing, but my intention was that the test files (which are there in testdata) should only "exist" if the test case def'n says they should, i.e. if the test case def'n includes them in `tt.locs`. The files in testdata are always there though, so os.Open on L397 would always work.

File: pkg/scorecard.go (hunk 1 above)
Comment (raghavkaul) [reply pkg/scorecard.go]: Fixed, only logging if annotation can't be parsed.

**Pull request merged**